---
title: "0010-parentage"
author: "Mac Campbell"
date: "August 11, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
```

## Aligning fastqs in data/concat

See 0010_Align.sh     

Missing AM6-O, AM6-P running 0010_Align.sh for these. Hmm... May need to make sure samplelist has a newline at the end?
Editing script and redoing.      

Don't forget to use -rf 0007/nonparalogous-contigs.tsv with ANGSD snp calling.

Now we have everything we want, but, we don't have a proper stats file.



```{r}
stats<-read_csv("0010/samplelist.stats", col_names = c("Reads","Align","DedupReads"))
samples<-read_tsv("0010/samplelist", col_names=c("Sample"))

#ARGGHHH doesn't simple delimit
comb<-bind_cols(samples, stats) 
#%>% separate(Sample, into=c("Source","Type"), sep="-", remove=FALSE)
```


There are some optional parents to include, but I don't know what are possible parents and which aren't. Sounds like some unknown parents are ok, I also need to get known fathers/mothers sorted.

```{r}
write_tsv(comb, "0010/parentage-read-counts.tsv")
highrollers<-comb %>% filter(DedupReads > 40000)
write_tsv(highrollers, file="0010/highrollers.txt")
```

```{r}
seq<-read_tsv("0010/toGenotype.txt")
```

Create a bamlist: 0010/concat_AM6-P_sorted_proper_rmdup.bam

```{r}
seq$File<-gsub(".stats","_sorted_proper_rmdup.bam", seq$Sample)
seq <- seq %>% mutate(Path=paste0("0010/",File))
write_tsv(select(seq,Path), "bamlists/parentage.bamlist",col_names = FALSE)
```


Following https://ryanpeek.org/radseq/pipeline_colony.html#outline:_the_bare_bones     

`~/bin/angsd -bam $list -GL 1 -out $output -doMaf 2 -minInd 20 -doMajorMinor 1 -SNP_pval 0.000001 -doGeno 4 -doPost 2 -postCutoff 0.95 -minMaf 0.005`

```{sh, eval=FALSE}
srun -p high -t 48:00:00 --nodes=2 $HOME/angsd/angsd -P 12 -bam bamlists/parentage.bamlist -GL 1 \
-out 0010/parentage -doMaf 2 -minInd 22 -rf 0007/nonparalogous-contigs.tsv \
-doMajorMinor 1 -SNP_pval 1e-2 -doGeno 4 -doPost 2 -postCutoff 0.95 -minMaf 0.005 -minMapQ 10 -minQ 20 > 0010/parentage.stdout 2> 0010/parentage.stderr &
```

We have 5017 snps.

Great, ryanpeek.org is down.    

gunzip parentage.geno.gz

Drawing 300 snps with a min maf of 0.25.
```{sh}
(base) maccamp@farm:~/spineflower/0010$ ../1000_scripts/Geno2Colony.pl  parentage.geno ../bamlists/parentage.bamlist > colony.input
 cat colony.input |  perl -pe 's/0010\/concat_//g' > colony-2.input
 head -n 4 colony.input > markers
 base) maccamp@farm:~/spineflower/0010$ grep "O_sorted" colony-2.input  >offspring
(base) maccamp@farm:~/spineflower/0010$ grep "P_sorted" colony-2.input  >parents
(
```

Need a header/footer
'spineflower'  !Dataset name
'spineflower'  !Output file name
12          ! Number of offspring in the sample
300           ! Number of loci
1234        ! Seed for random number generator
0           ! 0/1=Not updating/updating allele frequency
1           ! 2/1=Dioecious/Monoecious species
1           ! 0/1=No inbreeding/inbreeding
0           ! 0/1=Diploid species/HaploDiploid species
0  0        ! 0/1=Polygamy/Monogamy for males & females
0           ! 0/1=Clone inference =No/Yes
1           ! 0/1=Full sibship size scaling =No/Yes
1 1.0 1.0   ! 0,1,2,3=No,weak,medium,strong sibship size prior; mean paternal & meteral sibship size
1           ! 0/1=Unknown/Known population allele frequency

Footer: Can include known parentage relationships I guess.
printf '2 %.0s' {1..300} > allele.counts
Need the frequency of each allele per locus.... 
Wrote count.pl; now I need to get the alleles from the colony.dat file...

Working grep: grep "19111\s351"  parentage.geno     
(base) maccamp@farm:~/spineflower/0010$ head -n 1 colony.input | perl -pe 's/\s+/\n/g' | perl -pe 's/_/\\s/g' >> snps.txt
(base) maccamp@farm:~/spineflower/0010$ cat snps.txt | perl -pe 's/^(.)/grep "^$1/g' | perl -pe 's/(.)$/$1\\s" parentage.geno >> snps.geno/g' > snps-commands.sh

./count.pl snps.geno > tallies.txt
Turns out count.pl doesn't work when Major AF = Minor AF

  `(base) maccamp@farm:~/spineflower/0010$ cat header.txt space.txt allele.counts space.txt tallies.txt space.txt block.txt markers space.txt offspring probs space.txt parents space.txt parents space.txt options  > spineflower.dat`    

```{sh}
srun -p high -t 24:00:00 $HOME/colony/colony2s.ifort.out IFN:./spineflower.dat > colony-program.out 2> colony-program.stderr &
```

