---
title: "100-genome-align"
author: "Mac Campbell"
date: "8/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```


```{r}
library(tidyverse)
library(vcfR)
library(adegenet)
library(reshape2)
```

## Reference Genome 

How does our reference genome compare to our de novo assembly at this point?     

There is a chromosome-level assembly of Fagopyrum: GCA_002319775.1. Downloading and indexing in ~/genomes/fagopyrum     

I'll align the 187 pop gen samples to it and then see how snp numbers compare. We may need to relax proper pairing, so using picards mark duplicates to remove PCR duplicates without proper pairing..      

Not forgetting to remove the non-target species LG58, LG59, LG64, LG65, LG71, paralogs of course    

java -jar picard.jar MarkDuplicates \          
      I=input.bam \       
      O=marked_duplicates.bam \       
      M=marked_dup_metrics.txt      
      
```{sh, eval=FALSE}
(base) maccamp@farm:~/spineflower/data$ ~/gatk-4.1.9.0/gatk MarkDuplicates I=MC24.sort.bam O=temp.marked.bam M=temp.txt REMOVE_SEQUENCING_DUPLICATES=TRUE
```

(base) maccamp@farm:~/spineflower/data$ samtools flagstat MC24.sort.bam
429380 + 0 in total (QC-passed reads + QC-failed reads)
0 + 0 secondary
244 + 0 supplementary
0 + 0 duplicates
243594 + 0 mapped (56.73% : N/A)
429136 + 0 paired in sequencing
214568 + 0 read1
214568 + 0 read2
235090 + 0 properly paired (54.78% : N/A)
237532 + 0 with itself and mate mapped
5818 + 0 singletons (1.36% : N/A)
2314 + 0 with mate mapped to a different chr
936 + 0 with mate mapped to a different chr (mapQ>=5)
      
(base) maccamp@farm:~/spineflower/data$ samtools flagstat temp.marked.bam 
428660 + 0 in total (QC-passed reads + QC-failed reads)
0 + 0 secondary
244 + 0 supplementary
210096 + 0 duplicates
242874 + 0 mapped (56.66% : N/A)
428416 + 0 paired in sequencing
214208 + 0 read1
214208 + 0 read2
234378 + 0 properly paired (54.71% : N/A)
236812 + 0 with itself and mate mapped
5818 + 0 singletons (1.36% : N/A)
2306 + 0 with mate mapped to a different chr
930 + 0 with mate mapped to a different chr (mapQ>=5)

De novo assembly?

(base) maccamp@farm:~/spineflower/data$ samtools flagstat ../1004_alignments/MC24_sorted_proper_rmdup.bam
164667 + 0 in total (QC-passed reads + QC-failed reads)
0 + 0 secondary
124 + 0 supplementary
0 + 0 duplicates
164667 + 0 mapped (100.00% : N/A)
164543 + 0 paired in sequencing
82187 + 0 read1
82356 + 0 read2
164543 + 0 properly paired (100.00% : N/A)
164543 + 0 with itself and mate mapped
0 + 0 singletons (0.00% : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)

Yeah, so we have a lot more mapped reads to the genome. Let's see about the number snps.    

Loop it:

```{sh, eval=FALSE}
srun  -p high -t 6:00:00 ls | grep sort.bam | grep -v "bai" | while read line; do ~/gatk-4.1.9.0/gatk MarkDuplicates I=$line O="`basename $line .sort.bam`".dedup.bam M="`basename $line .sort.bam`".txt REMOVE_SEQUENCING_DUPLICATES=TRUE ; done;

```

Analyze, first by generating bamlist

```{r}
meta<-read_csv("meta/182.csv") %>% mutate(GenomeAlign=paste0("data/",Label,".dedup.bam"))
meta<-meta %>% separate(Location, into=c("Area","Site"), sep=",", remove=FALSE)
meta$Site<-trimws(meta$Site)
bamlist<-meta %>% select(GenomeAlign)

write_tsv(bamlist, "bamlists/182-genome.bamlist", col_names=FALSE)
```

Generate outputs. I don't think paralogs are going to do much, but will generate an SFS to see what we get.

```{sh, eval=FALSE}
##Generating SFS
srun -p high -t 48:00:00 --mem=16G --nodes=2 $HOME/angsd/angsd -P 12 -bam bamlists/182-genome.bamlist -doSaf 1 -out outputs/100/with-paralogs -anc $HOME/genomes/fagopyrum/GCA_002319775.1_Ft1.0_genomic.fna -GL 2 -minMapQ 10 -minQ 20 > outputs/100/with-paralogs-saf.stdout 2> outputs/100/with-paralogs-saf.stderr &

srun  -p bigmemh --mem-per-cpu=10G --mem=32G -t 12:00:00 $HOME/angsd/misc/realSFS outputs/100/with-paralogs.saf.idx -maxIter 100 -P 1 -fold 1 >  outputs/100/with-paralogs.sfs 2> outputs/100/realSFS.stderr


##Generating Beagle file

#Wow! 577 sites with -minInd 164 and -SNP_pval 1e-6
#Trying -minInd 137 and -SNP_pval 1e-2, 828 sites
#This worked when bam files weren't indexed (!) Trying again after indexing *.dedup.bam files
srun  -p bigmemh --mem-per-cpu=10G --mem=32G -t 12:00:00 $HOME/angsd/angsd -P 4 -bam bamlists/182-genome.bamlist -out outputs/100/pca -minInd 137 -GL 1 -doGlf 2  -doMajorMinor 1 -doMaf 2 -SNP_pval 1e-2 -minMapQ 10 -minQ 20  > outputs/100/pca.stdout 2> outputs/100/pca.stderr &

#362 used with pca the first time, 465 the second.
srun -p high -t 01:00:00 python $HOME/pcangsd/pcangsd.py -beagle outputs/100/pca.beagle.gz -kinship -admix -o outputs/100/pca-kin -threads 10



srun -p high -t 01:00:00 python $HOME/pcangsd/pcangsd.py -beagle outputs/100/pca.beagle.gz -relate outputs/100/pca-kin.kinship.npy -admix -o outputs/100/pca-relate -threads 10
```


### PCA

```{r}
cov<-read_delim("outputs/100/pca-kin.cov", col_names=FALSE, delim=" ") %>% as.matrix()
```

```{r}
#' @param samples character vector with the individuals IDs in the order in which
#' they were passed in the bamlist to angsd.
#' @param cov covariance matrix
covar2pcs <- function(samples, cov) {
  
  
  eig <- eigen(cov, symm = TRUE)
  PC <- as.data.frame(eig$vectors) %>%
    as_tibble() %>%
    setNames(sprintf("PC-%02d", 1:ncol(.)))
  
  samtib <- tibble(sample = samples)
  
  list(
    PCs = bind_cols(samtib, PC),
    eigevalues = eig$values
  )
}
```

```{r}
pca <- covar2pcs(meta$Label, cov)

pca_long <- pca$PCs %>%
  tidyr::gather(., key = "PC", "val", -sample)

# then expand a grid of the possible comparisons (ordered)
expg <- expand.grid(sample = pca$PCs$sample,
                    PCx = sprintf("PC-%02d", 1:6),
                    PCy = sprintf("PC-%02d", 1:6),
                    stringsAsFactors = FALSE) %>%
  tibble::as_tibble()

# then left join the pca results onto that
pca_pairs <- dplyr::left_join(expg, pca_long, by = c("sample", "PCx" = "PC")) %>%
  dplyr::rename(val_x = val) %>%
  dplyr::left_join(pca_long, by = c("sample", "PCy" = "PC")) %>%
  dplyr::rename(val_y = val)

npc <- 3
pc <- pca_pairs %>%
  filter( (PCx %in% sprintf("PC-%02d", 1:npc)) & 
            (PCy %in% sprintf("PC-%02d", 1:npc)) ) %>%
  left_join(., meta, by = c("sample" = "Label")) %>%
  mutate(group = Location) 

eig <- eigen(cov, symm = TRUE)
var <-eig$values/sum(eig$values)
cumvar <-cumsum(eig$values)/sum(eig$values)

head(var)
head(cumvar) 
```

```{r}
ggplot(pc) +
  geom_point(aes(x=val_x, y=val_y, fill=Site, shape=Area), alpha=0.75) +
  theme_bw() +
  scale_shape_manual(values = c(21,24)) +
  facet_grid(PCx ~ PCy) +
  guides(fill = guide_legend(override.aes=list(shape=22))) 

ggsave("outputs/100/pca.pdf", width=10, height=8)
```


Now let's up the SNP calls with freebayes. Or not.

````{sh, eval=FALSE}

srun -p bigmemh --mem-per-cpu=10G --mem=32G -t 12:00:00 freebayes-parallel <(fasta_generate_regions.py $HOME/genomes/fagopyrum/GCA_002319775.1_Ft1.0_genomic.fna 100000) 12  \
-f $HOME/genomes/fagopyrum/GCA_002319775.1_Ft1.0_genomic.fna -L bamlists/182-genome.bamlist > outputs/100/freebayes.vcf 2> outputs/100/freebayes.err &

#wants to write to unaccessible directory? cat: /dev/fd/63: No such file or directory
#Using non parallel version
srun -p bigmemh --mem-per-cpu=10G --mem=32G -t 12:00:00 freebayes -f  $HOME/genomes/fagopyrum/GCA_002319775.1_Ft1.0_genomic.fna \
--region CM008279.1  -L bamlists/182-genome.bamlist > outputs/100/freebayes.vcf 2> outputs/100/freebayes.err &
#4564 variants (including indels unfiltered across chrom1)
```
Have 28957 variants, but only one individual. So, we need to come up with some read groups.

```{sh, eval=FALSE}
srun  -p high -t 6:00:00 ls | grep dedup.bam | grep -v "bai" | while read line; do samtools addreplacerg $line -r ID:"`basename $line .dedup.bam`" | samtools view - -b > "`basename $line .dedup.bam`".rg.bam; samtools index "`basename $line .dedup.bam`".rg.bam; done;

 ls | grep rg.bam > rgs.bamlist
  cat rgs.bamlist | while read line; do echo "data/"$line >> rgs.path.bamlist; done;
  
srun -p bigmemh --mem-per-cpu=10G --mem=32G -t 12:00:00 freebayes -f  $HOME/genomes/fagopyrum/GCA_002319775.1_Ft1.0_genomic.fna \
 -L data/rgs.path.bamlist > outputs/100/freebayes.vcf 2> outputs/100/freebayes.err &
```

Malformed read group header!
Sigh.     

```{sh}
vcftools --vcf outputs/100/freebayes.vcf --remove-indels --max-alleles 2  --maf 0.05 --max-missing .1 --recode --out outputs/100/freebayes-filtered-90
```

1578 Sites @ 75% 
1578 Sites at 90%

Let's take a look:

```{sh, eval=FALSE}
cut -f 2 -d ',' meta/182.csv | perl -pe 's/Label\n//g' > outputs/100/samples.txt 
bcftools reheader --samples samples.txt -o outputs/100/recode.reheadered.vcf outputs/100/freebayes-filtered-90.recode.vcf
```

```{r}
meta182<-read_csv("meta/182.csv")
```

```{r, eval=FALSE}
vcf<-read.vcfR(file="outputs/100/freebayes-filtered-90.recode.vcf")
genind<-vcfR2genind(vcf)
genind@pop<-as.factor(meta$Location)
save(genind, file="outputs/100/genind.rda")
```
## Wayward samples

I need to find a bunch of samples that have to do with parentage. Plates BMAG058 and BMAG60    

Notes in 1001_sequence/plates/info.txt  BMAG058 is a concatenation of two seperate runs, 58 and 60
AM14-O
AM14-P
AM6-O
AM6-P
GV5-O
GV5-P
Lion12-P
Lion7-O
LM1-10-O
LM1-10-P
LM1-22-O
LM1-22-P
LM2-8-O
LM2-8-P
LM3-12-O
LM3-12-P
LM3-16-O
LM3-16-P
LM4-14-O
LM4-14-P
LM4-16-O
LM4-16-P
LM5-13-O
LM5-13-P
LM5-4-O
LM5-4-P
LM6-9-O
LM6-9-P
Long3-O
Long3-P
POT1-O
POT1-P
SMG2-O
SMG2-P
VCC2-O
VCC2-P

What about in here?
/share/schreierlab/Spineflower/ANGSD_analysis/BMAG060, probably on Berbera

Found samples used in colony:
concat_AM14-O_RA.sort.proper.rmdup.bam
concat_AM14-P_RA.sort.proper.rmdup.bam
concat_AM6-O_RA.sort.proper.rmdup.bam
concat_AM6-P_RA.sort.proper.rmdup.bam
concat_GV5-O_RA.sort.proper.rmdup.bam
concat_GV5-P_RA.sort.proper.rmdup.bam
concat_Lion12-P_RA.sort.proper.rmdup.bam
concat_Lion7-O_RA.sort.proper.rmdup.bam
concat_LM1-10-O_RA.sort.proper.rmdup.bam
concat_LM1-10-P_RA.sort.proper.rmdup.bam
concat_LM1-22-O_RA.sort.proper.rmdup.bam
concat_LM1-22-P_RA.sort.proper.rmdup.bam
concat_LM2-8-O_RA.sort.proper.rmdup.bam
concat_LM3-12-P_RA.sort.proper.rmdup.bam
concat_LM3-16-O_RA.sort.proper.rmdup.bam
concat_LM4-14-O_RA.sort.proper.rmdup.bam
concat_LM4-16-O_RA.sort.proper.rmdup.bam
concat_LM5-13-O_RA.sort.proper.rmdup.bam
concat_LM5-13-P_RA.sort.proper.rmdup.bam
concat_LM5-4-O_RA.sort.proper.rmdup.bam
concat_LM6-9-O_RA.sort.proper.rmdup.bam
concat_LM6-9-P_RA.sort.proper.rmdup.bam
concat_POT1-O_RA.sort.proper.rmdup.bam
concat_POT1-P_RA.sort.proper.rmdup.bam
concat_SMG2-O_RA.sort.proper.rmdup.bam
concat_SMG2-P_RA.sort.proper.rmdup.bam

Now in data/concat
for f in temp/*fastq; do mv -- "$f" "${f%.fastq}.fq"; done;
